crowbar_framework/app/models/crowbar_service.rb

```diff
@@ -338,10 +339,21 @@ def prepare_nodes_for_os_upgrade
       end
       if ready_for_reboot
         @logger.debug("Rebooting node #{node.name} for operating system upgrade")
-        node.ssh_cmd("/sbin/reboot")
+        ssh_status = node.ssh_cmd("/sbin/reboot")
+        if ssh_status[0] != 200
+          @logger.error("Upgrade failed for machine #{node.name}. Could not ssh.")
+          upgrade_nodes_failed.push(node.name)
+        end
+      else
+        @logger.error("Upgrade failed for #{node.name}. Node not ready for reboot")
+        upgrade_nodes_failed.push(node.name)
       end
     end
-    true
+    if upgrade_nodes_failed.any?
+      return false, upgrade_nodes_failed
+    else
+      return true
```

Comment:
Actually I found this quite dangerous because the `false` results is always evaluated as a `true` in a Boolean context. IMHO is better to return, for this method, always an array and force the unpack in every call.